src/backend/src/services/slack.services.ts

```diff
@@ -0,0 +1,179 @@
+ */
+export interface SlackMessage extends SlackMessageEvent {
+  user: string;
+  client_msg_id: string;
+  text: string;
+  blocks: {
+    type: string;
+    block_id: string;
+    elements: any[];
+  }[];
+}
+
+/**
+ * Represents a slack message event for a deleted message.
+ */
+export interface SlackDeletedMessage extends SlackMessageEvent {
+  subtype: 'message_deleted';
+  previous_message: SlackMessage;
+}
+
+/**
+ * Represents a slack message event for an edited message.
+ */
+export interface SlackUpdatedMessage extends SlackMessageEvent {
+  subtype: 'message_changed';
+  message: SlackMessage;
+  previous_message: SlackMessage;
+}
+
+/**
+ * Represents a block of information within a message. These blocks with an array
+ * make up all the information needed to represent the content of a message.
+ */
+export interface SlackRichTextBlock {
+  type: 'broadcast' | 'color' | 'channel' | 'date' | 'emoji' | 'link' | 'text' | 'user' | 'usergroup';
+  range?: string;
+  value?: string;
+  channel_id?: string;
+  timestamp?: number;
+  name?: string;
+  unicode?: string;
+  url?: string;
+  text?: string;
+  user_id?: string;
+  usergroup_id?: string;
+}
+
+export default class slackServices {
+  /**
+   * Given a slack event representing a message in a channel,
+   * make the appropriate announcement change in prisma.
+   * @param event the slack event that will be processed
+   * @param organizationId the id of the organization represented by the slack api
+   * @returns an annoucement if an announcement was processed and created/modified/deleted
+   */
+  static async processMessageSent(event: SlackMessageEvent, organizationId: string): Promise<Announcement | undefined> {
+    //get the name of the channel from the slack api
+    const slackChannelName: string = (await getChannelName(event.channel)) ?? `Unknown_Channel:${event.channel}`;
+    const dateCreated = new Date(1000 * Number(event.event_ts));
+
+    //get the message that will be processed either as the event or within a subtype
+    let eventMessage: SlackMessage;
+
+    if (event.subtype) {
+      switch (event.subtype) {
+        case 'message_deleted':
+          //delete the message using the client_msg_id
+          eventMessage = (event as SlackDeletedMessage).previous_message;
+          return AnnouncementService.deleteAnnouncement(eventMessage.client_msg_id, organizationId);
+        case 'message_changed':
+          eventMessage = (event as SlackUpdatedMessage).message;
+          break;
+        default:
+          //other events that do not effect announcements
+          return;
+      }
+    } else {
+      eventMessage = event as SlackMessage;
+    }
+
+    //loop through the blocks of the meta data while accumulating the
+    //text and users notified
+    let messageText = '';
+    let userIdsToNotify: string[] = [];
+
+    //Get the settings of all users in this organization to compare slack ids
+    const users = await UsersService.getAllUsers();
```

Comment:
this is getting all of the users in all organizations currently. IMO we should move this entire section into the getUserName function and make a raw prisma query for userSettings which would be more efficient here instead of having to make individual queries for all the users. would also shrink this function a bit which would be nice

Reply:
Ok, I can move the logic and prisma queries to work like that. I'm gonna put it into a utils function that gets a user id given a slack id just because then it reduces the amount of work the slack integration function is doing which can't be tested